component/s3/file/s3.go

```diff
@@ -0,0 +1,223 @@
+		return nil
+	}
+	s3Client := s3.NewFromConfig(*s3cfg, func(s3o *s3.Options) {
+		s3o.UsePathStyle = newArgs.Options.UsePathStyle
+	})
+
+	bucket, file := getPathBucketAndFile(newArgs.Path)
+
+	s.mut.Lock()
+	defer s.mut.Unlock()
+	s.args = newArgs
+	s.watcher.updateValues(bucket, file, newArgs.PollFrequency, s3Client)
+
+	return nil
+}
+
+// CurrentHealth returns the health of the component
+func (s *S3) CurrentHealth() component.Health {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+	return *s.health
+}
+
+func generateS3Config(args Arguments) (*aws.Config, error) {
+	configOptions := make([]func(*aws_config.LoadOptions) error, 0)
+	// Override the endpoint
+	if args.Options.Endpoint != "" {
+		endFunc := aws.EndpointResolverWithOptionsFunc(func(service, region string, _ ...interface{}) (aws.Endpoint, error) {
+			return aws.Endpoint{URL: args.Options.Endpoint}, nil
+		})
+		endResolver := aws_config.WithEndpointResolverWithOptions(endFunc)
+		configOptions = append(configOptions, endResolver)
+	}
+
+	// This incredibly nested option turns off ssl
+	if args.Options.DisableSSL {
+		httpOverride := aws_config.WithHTTPClient(
+			&http.Client{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{
+						InsecureSkipVerify: args.Options.DisableSSL,
+					},
+				},
+			},
+		)
+		configOptions = append(configOptions, httpOverride)
+	}
+
+	// check credentials
+	if args.Options.AccessKey != "" {
+		if args.Options.Secret == "" {
+			return nil, fmt.Errorf("if accesskey or secret are specified then the other must also be specified")
+		}
+		credFunc := aws.CredentialsProviderFunc(func(ctx context.Context) (aws.Credentials, error) {
+			return aws.Credentials{
+				AccessKeyID:     args.Options.AccessKey,
+				SecretAccessKey: string(args.Options.Secret),
+			}, nil
+		})
+		credProvider := aws_config.WithCredentialsProvider(credFunc)
+		configOptions = append(configOptions, credProvider)
+	}
+
+	cfg, err := aws_config.LoadDefaultConfig(context.TODO(), configOptions...)
+	if err != nil {
+		return nil, err
+	}
+	// Set region
+	if args.Options.Region != "" {
+		cfg.Region = args.Options.Region
+	}
+
+	return &cfg, nil
+}
+
+// handleContentUpdate reads from the update and error channels setting as appropriate
+func (s *S3) handleContentUpdate(ctx context.Context) {
+	for {
+		select {
+		case buf := <-s.updateChan:
+			s.mut.Lock()
+			strBuf := string(buf)
+			// Only update if changed
+			if strBuf != s.content {
+				s.content = strBuf
+				s.handleContentPolling(nil)
+			}
+			s.mut.Unlock()
+		case err := <-s.errorChan:
+			s.mut.Lock()
+			s.handleContentPolling(err)
+			s.mut.Unlock()
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (s *S3) handleContentPolling(err error) {
+	if err == nil {
+		s.opts.OnStateChange(Exports{
+			Content: rivertypes.OptionalSecret{
+				IsSecret: s.args.IsSecret,
+				Value:    s.content,
+			},
+		})
+		s.health.Health = component.HealthTypeHealthy
+		s.health.Message = "s3 file updated"
```

Comment:
You're going to need to protect s.health with a mutex since CurrentHealth can be called at any time.

Reply:
Should be handled by the caller but moving the mutexes around to make it clearer.